File: server/src/routes/userRoutes.ts
import { Elysia, type Context } from "elysia";
import * as auth from "../lib/auth";

function userRoutes(app: Elysia) {
	app.post("/api/v1/login", async (cxt: Context) => {
		return await userLogin(cxt);
	})

	app.post("/api/v1/register", async (req) => {
		return await userRegister(req.body);
	})
}


async function userRegister(ctx: Context): Promise<string> {

    const body = await ctx.request.json;
	if (!body) {
		return "Empty body"
	}
	if (!auth.isRegistrationInput) {
		return "Invalid input: requires body properties 'username', 'email', and 'password' as string";
	}
	return await auth.registerUser(
		body.username,
		body.email,
		body.password
	);
}

async function parseBody<T>(request: Request): Promise<T> {
    const body = await request.json();
    return body as T;
}

async function userLogin(ctx: Context): Promise<string> {
        const body = ctx.request.body;
		if (!body) {
			return "Empty body";
		}
		if (!auth.isLoginInput) {
			return "Invalid input: requires body properties 'identifier' and 'password' as string";
		}

		const { identifier, password }: auth.LoginInput = body as auth.LoginInput;

		try {
			const authToken = await auth.loginUser({identifier: identifier, password: password});
			return authToken;
		} catch (err) {
			if (err instanceof auth.IncorrectPasswordError) {
				return err.message;
			}
			else if (err instanceof auth.UserNotFoundError) {
				return err.message;
			}
			else if (err instanceof auth.PrivateKeyNotFound) {
                res.status(500).send(err.message);
				return err.message;
			}
			else if (err instanceof auth.PublicKeyNotFound) {
                res.status(500).send(err.message);
                return err.message;
			}
			else {
				console.log(`Unknown error unable to login ${err}`);
				return `Unknown error unable to login ${err}`;
			}
		}
}

export default userRoutes;

File: server/src/routes/user_routes.rs
use axum::{
    extract::{Json, Extension},
    routing::post,
    Router,
    response::Response,
    http::StatusCode,
    body::Body
};
use serde_json::json;
use serde::Deserialize;
use tokio_postgres::Client;
use std::sync::Arc;

use crate::User;
use crate::auth::login;
use crate::auth::register;

#[derive(Debug, Deserialize)]
struct LoginInput {
    identifier: String,
    password: String
}

#[derive(Debug, Deserialize)]
struct RegistrationInput {
    email: String,
    username: String,
    password: String
}

pub fn user_routes() -> Router {
    Router::new()
        .route("/api/v1/login",post(post_login))
        .route("/api/v1/register",post(post_register))
}

async fn post_login(Extension(client): Extension<Arc<Client>>, Json(payload): Json<LoginInput>) -> Response {
    match login::login(&client, &payload.identifier, &payload.password).await {
        Ok(token) => {
            let response_body: serde_json::Value = json!({
                "token": token
            });

            Response::builder()
                .status(StatusCode::OK)
                .header("Content-Type", "application/json")
                .body(Body::from(response_body.to_string()))
                .unwrap()
        },
        Err(err) => {
            Response::builder()
                .status(StatusCode::BAD_REQUEST)
                .header("Content-Type", "application/json")
                .body(Body::from(err.to_string()))
                .unwrap()
        },
    }
}

async fn post_register(Extension(client): Extension<Arc<Client>>, Json(payload): Json<RegistrationInput>) -> Response {
    let user: User = match register::register(&client, &payload.username, &payload.email, &payload.password).await {
        Ok(u) => u,
        Err(err) => {
            return Response::builder()
                .status(StatusCode::BAD_REQUEST)
                .header("Content-Type", "application/json")
                .body(Body::from(err.to_string()))
                .unwrap();
        },
    };

    match login::login(&client, &user.username, &user.password).await {
        Ok(token) => {
            let response_body: serde_json::Value = json!({
                "token": token
            });

            Response::builder()
                .status(StatusCode::OK)
                .header("Content-Type", "application/json")
                .body(Body::from(response_body.to_string()))
                .unwrap()
        },
        Err(err) => {
            Response::builder()
                .status(StatusCode::BAD_REQUEST)
                .header("Content-Type", "application/json")
                .body(Body::from(err.to_string()))
                .unwrap()
        },
    }
}
